### ISSUE TITLE
Allow to pin Tabs (sort) - Configure Map Tracks

### ISSUE THREAD

**Issue by Rowin63** [osmandapp/OsmAnd#16663]
### Description
The tabs in the new "configure map / tracks"-menu are not in alphabetical order, but have been before.
### Your Environment
OsmAnd Version:  4.4.0#14226
Android/iOS version: 
Device model:

**Comment by Zirochkabila** [comment]
@Rowin63 You can sort them in any order you want
<table style="width:100%">
  <tr>
<th><img src="https://user-images.githubusercontent.com/104760013/224480014-aa9c1377-2bd8-478d-a0f3-405db7e0f590.jpeg"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by Rowin63** [comment]
Yes, the tracks. But not the tabs (folder)

**Comment by Zirochkabila** [comment]
I agree
Perhaps we should add a tab sorting function (something similar to how it is implemented with tracks)

**Comment by Rowin63** [comment]
It was in alphabetic order one of the last releases

**Comment by sonora** [comment]
If @Rowin63 is right, Ivwould consider this an issue. But actually, so far on my devices I can't reproduce. But the scren crashes under some circumstances, so maybe I ca not get to the situation to reproduce (https://github.com/osmandapp/OsmAnd/issues/14890#issuecomment-1462886792)...

**Comment by Rowin63** [comment]
In 4.4.1 tabs are alphabetical.
In 4.4.0#14230 and the one before tabs are not alphabetical anymore (neither the overview page nor the tabs themselves)

**Comment by sonora** [comment]
Build 14230 is newer, so I'd say this is an Issue, well spotted!

In particular also because of the following situation: If you have your GPX tracks organized in subfolders like `<osmand-data-folder>/tracks/Skitours/Variants`, our legacy tracks screen under MyPlaces displays these folders as
- Skitours
- Skitours/Variants

but this new screen (both in its folder ribbon and on the "Switch folder" dialog) just displays
- Skitours
- Variants

i.e. suppresses any connection to the folder hierarchy. You cannot see which parent folder the "Variants" belong to (you may have several Variants folders belonging to different parents), unless the sorting is such that ithe parent is reliably the folder displayed immediately above / to the left.

**Comment by sonora** [comment]
PS: @Zirochkabila , @dmpr0 Not sure we need a "Folder sorting Option", I guess we would riduculing ourselves with not non-essential functionality  while we still miss a lot of essential functionality, like being able to search for tracks, get a single track displayed on thd map, fix the crashes, etc., see #14890 and #16508.

**Comment by Rowin63** [comment]
Its nice and helpful sorting the tracks, but no need for the folders. 
But please keep them in alphabetical order. Thanks!

**Comment by sonora** [comment]
@Rowin Can you please check the folder sorting with the brand new builds to see if you still have the issue? We had some fixes in the sorting code  and with successive builds I (still) cannot reproduce.

**Comment by Rowin63** [comment]
Hello!

Actual +Version is 4.4.0#14233 from today.
Still I can not see an option for sorting the folders. 
There is one option to change the folder (1)
And another option to change the appearance of the selected 0tracks (2)
But the folders still are mixed up.

 
![Tracks Menu](https://user-images.githubusercontent.com/95540936/224554328-a1b74671-f3f2-4486-a670-91a68693baf1.JPG)

**Comment by sonora** [comment]
Weird. Perhaps this was caused by some interim buggy build.

Do you have access to the tracks folder via some file manager app or a PC? It may be worth to exit OsmAnd, rename the folder to tracks0, restart OsmAnd, go to its tracks screen so OsmAnd realizes there now are zero tracks. Then kill OsmAnd again, rename tracks0 back (replacing the newly created, empty tracks folder). Does that fix it? Now it would be interesting if the issue comes back somehow. (and no, there is no folder sorting button).

**Comment by Rowin63** [comment]
I did as you wrote.
After renaming my original track folder back to "tracks",  the tabs/folders appear in exactly (!) the same mixed order like before! But their order is not by name and not by timestamp.

**Comment by sonora** [comment]
Maybe @Chumva could take a look. I have not investigated where we perform the folder sorting.

And it's weird that I cannot reproduce, for me the folders are sorted ok A-Z, both the order of the tabs in the ribbon and on the "Switch folder" dialog. (Other than the folder displayed as "Tracks", which is actually the root of the tracks folder and sorted to the top, probably ok.)

**Comment by dmpr0** [comment]
The tabs on this screen should only display top-level folders. It will be fixed in the main task.

**Comment by sonora** [comment]
@Rowin63 Are your folders shown above nested subfolders of each other (which would explain the phenomenon as per https://github.com/osmandapp/OsmAnd/issues/16663#issuecomment-1464894792 ),or are they all immediate subfolders of "tracks"?

**Comment by Rowin63** [comment]
I dont use subfolders at all..
All my track-folders are immediate (sub)folders of "tracks".

On two "Galaxy S20+" I use the same synced data in 4.4.1(+) and in 4.4.0#14233(~)
(+) is alphabetical
(~) is mixed up

![Screenshot_20230313_095005_OsmAnd+](https://user-images.githubusercontent.com/95540936/224652575-794d4afa-059e-4906-95b6-288879db6f31.jpg)
![Screenshot_20230313_094940_OsmAnd~](https://user-images.githubusercontent.com/95540936/224652640-6afe68b8-6341-4ba7-bee2-cb0d93d7852a.jpg)

**Comment by sonora** [comment]
So in that case IMHO we are looking at a true issue here not yet understood, which was closed prematurely, allow me to re-open.

**Comment by Rowin63** [comment]
I cant switch back to an earlier nightly release to find out. I dont remember anymore exactly in which release this appeared, must have been in one of the last builds before #14226. But sure it was not from the beginning, when you started with this new track menu. Then all items were in a correct order and not mixed up.

**Comment by vshcherb** [comment]
Currently tabs are sorted by last modified date which was before default, so it should stay as before, indeed we need to introduce Folder Sorting Name or Time cause this is more convenient then alphabetical

**Comment by vshcherb** [comment]
For backward compatibility and previous release tab order should be the same as in My Places cause this screen will remain untouched for now and it follows the sorting selected.

**Comment by Rowin63** [comment]
> ...indeed we need to introduce Folder Sorting _Name or Time_ cause this is more convenient then alphabetical

Sorry, I dont unterstand this...?
Name is more convenient than alphabetical? Whats the difference?

In the "previous" my-places-menu track-folders are always sorted A-Z or Z-A. 
Is this name or alphabetical? 

I use 40+ track-folders and I would find it very laborious to search for a folder named "A..." and find it somewhere in between or at the end of all others.

Do I misunderstand something?

**Comment by vshcherb** [comment]
Nope not always, if go to My Places it's sorted as you select it. And here it should follow the same sorting.

**Comment by Rowin63** [comment]
Ups I see, there is also an option "last edited". Never used it...
Everything will be fine with these three options in the new menu 👍

**Comment by sonora** [comment]
CanI suggest that
- we make the sorting criterion global, i.e. whatever you apply is used for folders/tabs, as well as for the tracks on each tab. That's how it is in the legacy screen, has worked well. And that's how most file manager apps jandle it.
- It saves us having yet another button or interface, and potential corresponding more mouse clicks. Also, the sorting is instantaneous now, so it's a _one click / one second_ action to change whenever needed?
- Regarding the subfolders, also here I think it is madatory for clarity that we do like on the old screen: We display the entire sub-path below the tracks folder, not just the lowest folder name?

**Comment by vshcherb** [comment]
You can suggest, but now it's different we used to have global sorting tracks / folders but now it's separate.
- Sorting in each folder will be preserved
- Sorting will be saved globally between sessions and screens My Places / Configure Map (will be fixed)
- Until we reintroduce folder sorting (present on My Places), tabs will follow sorting on My Places (will be fixed)
- For subfolders (for now, path will be displayed in the categories screen, but won't be in the tbas)

**Comment by sonora** [comment]
I wonder how "_path will be displayed in the categories screen, but won't be in the tabs_" can be usable for any sorting other than "A-Z", where at least you can conclude that the parent folder is the tab immediately to the left...

For all I can think of right now, the way around this could be to abandon the "_scrollable tab ribbon_" for a "_tappable path widget_" like in the mockup below: That seems superior both in fuctionality and handling?

![Screenshot_20230313-200117_My Files](https://user-images.githubusercontent.com/845735/224803514-94f1462e-b719-459d-9639-106f1153c601.jpg)

**Comment by Rowin63** [comment]
If I edit the color setting of a track in this new menu, all other, _previously untouched_ tracks, do also get this settings.
It seems, that my editing always applies to all "untouched" tracks.
The ones I edited before are not affected.

**Comment by Rowin63** [comment]
And please, try to find a simple solution how to work with this menu.

Lets say, I have three folders, every one includes 50 tracks. In each of these folders I want to edit 10 tracks, every 10 in with a specified color. How to proceed? It seems a bit complicated.

1) I have to deselect all actual displayed tracks.
2) go to first folder and select the 10 out of 50, edit them and apply
3) have to deselect all actual displayed tracks again.
4) go to second folder and select the 10 out of 50, edit them and apply
5) have to deselect all actual displayed tracks again.
6) go to third folder and select the 10 out of 50, edit them and apply.
7) now I have to go to folder one and two again and have to check every single track I´ve edited right now, one by one to display them.
8) And now, if now I want to edit other tracks in a fourth color, I have to startover from the beginning going through all the selecting procedure again, one by one.

At least I miss a "select all" checkbox in a folder.
In the "previous displayed"-tab I only can see the tracks from the very last session (it´s not as extensive as the regular "track history")

**Comment by dmpr0** [comment]
@Rowin63 this menu is not for described task. You can do it in new My Places – Tracks menu.
First and main task of this screen is to show list of tracks visible on the map, and ability to turn on more or disable visible.

**Comment by Rowin63** [comment]
I can´t do this editing steps in the "my places"-menu. There I only can switch on and off selected tracks.
To display tracks on a map (turn on and off) I really don´t need this new menu. This is already very easy in the "my places".

Your first intetion for this new menu was "_editing more than one track in one go_". Have to seek for this post here.

**Comment by Rowin63** [comment]
sonora noted: "it currently seems like we are expanding Configure map > Tracks to be a powerful configuration screen?"
vshcherb answered: "yes exactly"

This sounds more than just switching on and off.

**Comment by sonora** [comment]
Absolutely no need to have and maintain 2 screens in the long run. Every file manager logic can fulfill both purposes, mark one/several/no tracks, and configure/open/manipulate them from one control logic.

**Comment by dmpr0** [comment]
> I can´t do this editing steps in the "my places"-menu. There I only can switch on and off selected tracks. If I want to edit tracks, I have to use "configure map" and am forwarded to the new menu. For displaying tracks on a map (simple turn on and off) I really don´t need this new menu. This is already very easy in the existing "my places".
> 
> Your first intention for this new menu was "_editing more than one track in one go_". Have to seek for this post here.

Yes, we didn't start to implement it.

**Comment by Rowin63** [comment]
As I understood by now, the "old" my places / tracks section is to be replaced by the "new" configure map / tracks menu.
Am I right or wrong? 
If it is not, then it´s useless for me spending my time with any beta of the "new" menu, just for switching on and off tracks.
As a user who really likes Osmand I need and want a possibility to edit the look of a bunch of tracks in one go, comfortable and logic.

**Comment by dmpr0** [comment]
> As a user who really likes Osmand I need and want a possibility to edit the look of a bunch of tracks in one go, comfortable and logic
As I mentioned early many times, we didn't start to implement it. But we have prepared task that just wait until developers start to work on it. Updated Tracks screen in My Places will include all needed actions.

> If it is not, then it´s useless for me spending my time with any beta of the "new" menu, just for switching on and off tracks.
Yes you are right, and you don't need to spend any time on this screen if you want to edit favorites or upload edits to OSM.

**Comment by sonora** [comment]
@dmpr0

Maybe I misunderstand and hope it is due to some language barrier, but it seems you always try to confront us with "set decisions just not yet implemented", while we make suggestions how a concept could look that would potentially be significantly more useful and favorable?

Why is it that we cannot have an open discussion about the improvement ideas people develop and post, based on objective reasoning, without the knockout "_It will not be like this._" ??

**Comment by Rowin63** [comment]
If I´m through all the procedure mentioned above, then I´m back in the new "configure map / tracks"-menu and I see, for some reason I missed to cofigure the appearance of only one track (like in the screenshot), I can not do this from here where I am now, because any change I do would apply to all 158 previous (different) configured and displayed tracks (and I dont want to deselect them again, it was already a long "selecting" way until here).

I have to exit this menu, go back to "my places" and here I can edit the appearance of this single track. So far so good, of course it´s a way to do it. Now you give me two different menues, which basicly do very similar things, like track displaying and editing track appearance.

But: 
) in one menu (my places) I don´t have any idea of the appearance of a group of tracks, but I can easily edit a _single one_.
) in the other menue (configure map / tracks) you give me a very good overview of a tracks appearance, but I can´t edit (easily/comfortable) _a single one_ (like when my 158 other tracks are displayed).

Here in this new menu I still struggle a) with the "checking options" and b) with the fact, that editing always applies to all _displayed_ tracks (selected here = displayed). An option here _selected = editing_ (independend to the displayed tracks) would be a really great and helpful option. It would be so easy then, to configure different groups of tracks in different appearance and when finished, displaying them or not.

This is my intention from the beginning, when weeks ago I first started asking for an option to configure more than one track in one go. 

Configuring detached from displaying. This can be seen as the point of it. 
Just my opinion. Thanks.


![Screenshot_20230315_082113_OsmAnd+](https://user-images.githubusercontent.com/95540936/225239900-f6d7c10c-972e-4d1e-aed1-d8a07668e2ff.jpg)

**Comment by sonora** [comment]
And I hope the plan is not to introduce a _third screen_ for that use case... 😉😘😉 

Forgive me - just kidding. 😉

**Comment by dmpr0** [comment]
@sonora Hello! Thank you for the feedback. 
If you have any other questions please contact us at support@osmand.net

**Comment by vshcherb** [comment]
every file manager has 2 or even more modes: 1 to browse, 1 to select, 1 to paste item. The could seem very similar but actions and menus and behavior in each mode is different. So in the end it will be more clear what is different on selection screen / search / browse etc.

**Comment by vshcherb** [comment]
Tabs sorting could be very confusing cause tabs is a flat list but folders could be nested. So only 1 type of sorting could be non-confusing (A-Z / Z-A).

So we decide that most important for users is to pin individual folders and filters.

**Comment by vshcherb** [comment]
As smart folders will be displayed first it's possible they will be configurable and this task will rely on https://github.com/osmandapp/OsmAnd/issues/18221

**Comment by sonora** [comment]
I believe to have fixed this some time ago. The sort order of the _tabs_ on _Configure map > Tracks_ should now always be the same as the (selectable) sort order of the _folders_ on _MyPlaces > Tracks_.

**Comment by Rowin63** [comment]
I'm not sure.

4.8.0#2515:
"My places A-Z" is in "configure map A-Z", but Import, Rec and ÖTTT are listed at the end.
"My places Z-A" is in "configure map Z-A", but ÖTTT, Rec and Import are listet first.
"My places last edited" is completely different from "configure map last edited". 

And +4.7.17 is still confusing.
"My places A-Z" is in "configure map" still completely mixed up. 

I still don't understand this sorting at all, but I quit thinking about it.

**Comment by sonora** [comment]
We can re-open, but perhaps it is more promising to research the new situation and create a fresh issue.

The former problem was there was no controllable sorting at all for Configure map's tabs. Now it is coupled to the sorting of the folders as specified on MyTracks.

The phenomenon now may be that the tabs flatten the folders and theie subfolders into one list (it becomes visible when tapping the 'Switch folder'button to the left of the 3-dots). And that list creates some funny effects when the sorting is applied.Like for alphabetical sorting due to the fact that it keeps the ralative path names for the flattened folders (which is hidden on the tab headers). For "Last modified" sorting not sure,maybe it considers the Last modified date only for the first level, or for all levels.

It's rather complex, and I wish we had followed my early advice not to create two different Tracks screens with both differentt look & feel and different logic. Now we spend all this effort and life time to somehow align them in ways which make sense... 😞

**Comment by Rowin63** [comment]
You may be right with subfoders, but I don't use any.

**Comment by sonora** [comment]
Can you post some screenshots of obviously false sorting? Because other than the subfolder peculiarity I currently find no flaw... (and please note that I think my fixes are not in the officially publishe 4.7 17. Should use a current nightly build as a basis until they publish a 4.7.18 or a 4.8beta, or else we are hunting a ghost ... 😉

**Comment by Rowin63** [comment]
This is Screenshots from +4.7.17:

My places "A-Z" as well as "Z-A" is correct ( _Name, Import, ÖTTT and Rec are right in order, both sorting directions)

![Screenshot_20240618_194114_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/bf45cb3b-cd46-4011-af52-a94a1a475f92)

![Screenshot_20240618_194125_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/efaa89a3-0bab-4adf-a1e5-2a6c05b0a758)

![Screenshot_20240618_194134_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/09add5fd-64fc-4e02-8ad4-6fdcb866524a)

![Screenshot_20240618_194143_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/bdee995c-b0f9-42ea-85c9-77653fdd5d53)


But "Configure map / tracks" is comletely mixed up, and it´s also not "last edited".

![Screenshot_20240618_194201_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/afb5d4fa-3820-40c3-b775-52a5815723ca)

![Screenshot_20240618_194218_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/68dbe341-c8a1-4b0f-9aac-58efe4bbe6fb)

![Screenshot_20240618_194234_OsmAnd+_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/c7ed95ba-e58a-4b20-8642-e8f1ca7cc0a2)

**Comment by Rowin63** [comment]
This is screenshost from 4.8.0#2531

My places is correct in both directions (A-Z, Z-A).

And "configure map" is mainly also correct, except the folders _Name, Import, Rec and ÖTTT, they are listed at the end (A-Z) 

![Screenshot_20240618_195121_OsmAnd~_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/c506ab17-cb02-4c40-bfc9-2606ebd541a9)


or at the beginning (Z-A)

![Screenshot_20240618_195232_OsmAnd~_Bildgröße ändern](https://github.com/osmandapp/OsmAnd/assets/95540936/113d9466-b134-42d7-beaf-8046337c0457)

**Comment by sonora** [comment]
Yes, the uncontrollable 'randomness' you observe for Configure maps in 4.7.17 is exactly what I have fixed.

Regarding the 4.8 nightly, I have a feeling the sort algorithm as it is used for the tabs has am issue with special characters like 'ö' and perhaps also '_'. We can exüeriment a bit more to find out, and if confirmed it would support my suspicion that any remaining issue we have now with the tabs sorting is of a completely different nature than what was fixed before.

**Comment by Rowin63** [comment]
I agree for "_" and "Ö", but why "import" and "rec" ?

**Comment by sonora** [comment]
They are system-created folders, perhaps they have a special sorting treatment or something to not 'vanish' in a lengthy list of user-created folders.... but that's just a speculation on my behalf.

**Comment by Rowin63** [comment]
But only in "configure map" 😁
Ok, we'll see how this issue will be solved... 😉

**Comment by sonora** [comment]
BTW, I use a build of June 14 as my current OsmAnd 'workhorse', and that one sorts '_' correctly to the front before '_a , only the Umlaut 'ö' is inexplicably sorted to the very end, behind 'z'. And yes, only for the tabs, not for the folders.

Still recommend we research better, then open a fresh issue, this one has become too long for anybody to still distill the remaining issues from it...

**Comment by sonora** [comment]
PS: Basically see https://github.com/osmandapp/OsmAnd/issues/19634#issuecomment-2113169587.

I see #19634 as the master issue for this. There I have also linked my hotfix, and in deed I have just checked that I only catered for the order of tabs to facilitate setting A-Z and Z-A, the most basic ones. I had put a "PRELIMINARY" remark in my hotfix, and probably much of the issues we discussed above about current tab sorting would be fixed by someone streamling that hotfix to reuse the existing folder sorting code also for tabs.
